### PR TITLE
Add Debates engine

### PR DESCRIPTION
**Pull request by Dor3nz**
#### :tophat: What? Why?
Fixes https://github.com/AjuntamentdeBarcelona/decidim-barcelona/issues/3

### CODE REVIEW

File: engines/decidim-debates/app/models/decidim/debates/debate.rb [hunk 1]
```diff
@@ -23,6 +23,16 @@ def author_avatar_url
         author&.avatar&.url || ActionController::Base.helpers.asset_path("decidim/default-avatar.svg")
       end
 
+      # Public: Canpeople comment on this debate?
+      #
+      # Until we have a way to store options fore features and its resources we
+      # assume all debates can be commented.
+      #
+      # Returns Boolean
+      def commentable?
```
Comment (oriolgual) [inline engines/decidim-debates/app/models/decidim/debates/debate.rb]: Don't do it like this, use the settings like here: https://github.com/AjuntamentdeBarcelona/decidim/pull/638/files

Comment (Dor3nz) [inline engines/decidim-debates/app/models/decidim/debates/debate.rb]: I actually did both and forgot to remove this one :(

File: engines/decidim-debates/app/controllers/decidim/debates/debates_controller.rb [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+
+module Decidim
+  module Debates
+    # Exposes the debate resource so users can view them
+    class DebatesController < Decidim::Debates::ApplicationController
+      include FilterResource
+
+      helper_method :debates, :debate
+
+      def index; end
```
Comment (oriolgual) [inline engines/decidim-debates/app/controllers/decidim/debates/debates_controller.rb]: You don't need if there's a view named `index`

File: engines/decidim-debates/app/forms/decidim/debates/admin/debate_form.rb [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+        validates :title, translatable_presence: true
+        validates :description, translatable_presence: true
+        validates :instructions, translatable_presence: true
+        validates :start_time, presence: true, date: { before: :end_time }
+        validates :end_time, presence: true, date: { after: :start_time }
+
+        validates :current_feature, presence: true
+        validates :category, presence: true, if: ->(form) { form.decidim_category_id.present? }
+
+        def scope
```
Comment (oriolgual) [inline engines/decidim-debates/app/forms/decidim/debates/admin/debate_form.rb]: Should we validate the scope like category?

File: engines/decidim-debates/app/forms/decidim/debates/admin/debate_form.rb [hunk 4]
```diff
@@ -0,0 +1,37 @@
...
+        attribute :end_time, DateTime
+        attribute :decidim_category_id, Integer
+
+        validates :title, translatable_presence: true
+        validates :description, translatable_presence: true
+        validates :instructions, translatable_presence: true
+        validates :start_time, presence: true, date: { before: :end_time }
+        validates :end_time, presence: true, date: { after: :start_time }
+
+        validates :current_feature, presence: true
```
Comment (oriolgual) [inline engines/decidim-debates/app/forms/decidim/debates/admin/debate_form.rb]: The current feature is automatically injected, no need to validate.

File: engines/decidim-debates/app/models/decidim/debates/debate.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+      belongs_to :feature, foreign_key: "decidim_feature_id", class_name: Decidim::Feature
+      belongs_to :scope, foreign_key: "decidim_scope_id", class_name: Decidim::Scope
+      belongs_to :category, foreign_key: "decidim_category_id", class_name: Decidim::Category
+      has_one :organization, through: :feature
+
+      validate :category_belongs_to_organization
+      validate :author_belongs_to_organization
+      validates :title, presence: true
+
+      def author_name
```
Comment (oriolgual) [inline engines/decidim-debates/app/models/decidim/debates/debate.rb]: I'd remove the author name from here & the views. Debates are only created by admins so it makes no sense.

File: engines/decidim-debates/app/models/decidim/debates/debate.rb [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        user_group&.name || author&.name || I18n.t("decidim.debates.models.debate.fields.official_debate")
+      end
+
+      def author_avatar_url
```
Comment (oriolgual) [inline engines/decidim-debates/app/models/decidim/debates/debate.rb]: Ditto

File: engines/decidim-debates/app/views/decidim/debates/debates/index.html.erb [hunk 7]
```diff
@@ -0,0 +1,58 @@
...
+  <div class="row small-up-1 medium-up-2 large-up-3 card-grid">
+    <% debates.each do |debate| %>
+      <div class="column">
+        <article class="card card--debate">
+          <div class="card__content">
+            <div class="card__header">
+              <a href="/debate-view-AMA" class="card__link">
+                <h5 class="card__title"><%= translated_attribute debate.title %></h5>
+              </a>
+              <div class="card__author author-data author-data--small">
```
Comment (oriolgual) [inline engines/decidim-debates/app/views/decidim/debates/debates/index.html.erb]: Remove all author stuff

File: engines/decidim-debates/app/views/decidim/debates/debates/index.html.erb [hunk 8]
```diff
@@ -0,0 +1,58 @@
...
+                <div class="author-data__main">
+                  <div class="author author--inline">
+                    <a href="#" class="author__avatar author__avatar--small">
+                      <img src="<%= debate.author_avatar_url %>" alt="">
+                    </a>
+                    <a href="#" class="author__name">
+                      <%= debate.author_name %>
+                    </a>
+                    <%= l debate.created_at, format: "%d/%m/%Y" %>
```
Comment (oriolgual) [inline engines/decidim-debates/app/views/decidim/debates/debates/index.html.erb]: Remove this too.

File: engines/decidim-debates/app/views/decidim/debates/debates/show.html.erb [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+<% content_for :meta_title, debate.title %>
+<% content_for :meta_url, debate_url(debate.id) %>
+<% content_for :twitter_handler, current_organization.twitter_handler %>
+
+<div class="wrapper">
+  <div class="row column view-header">
+    <h2 class="heading2">
+      <%== translated_attribute debate.title %>
+    </h2>
+    <div class="author-data">
```
Comment (oriolgual) [inline engines/decidim-debates/app/views/decidim/debates/debates/show.html.erb]: Remove author stuff

File: engines/decidim-debates/app/views/decidim/debates/debates/show.html.erb [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+          </a>
+          <%= l debate.created_at, format: :long %>
+          <a href="#" class="author__name">
+            <%= debate.author_name %>
+          </a>
+        </div>
+      </div>
+      <div class="author-data__extra">
+        <a href="#comments" title="<%= t('.comments') %>">
+          <%= icon "comment-square", aria_label: t('.comments'), role: "img" %> <%= Decidim::Comments::Comment.where(commentable: debate).count %>
```
Comment (oriolgual) [inline engines/decidim-debates/app/views/decidim/debates/debates/show.html.erb]: I'd add a relation at debate, something like:

`has_many :comments, as: :commentable`

So you can just do `debate.comments.count`

/cc @josepjaume @beagleknight

File: engines/decidim-debates/spec/forms/debate_form_spec.rb [hunk 11]
```diff
@@ -1,3 +1,4 @@
+# frozen_string_literal: true
 # frozen_literal_string: true
```
Comment (mrcasals) [inline engines/decidim-debates/spec/forms/debate_form_spec.rb]: Remove this one

File: db/migrate/20170127093544_create_debates.decidim_debates.rb [hunk 12]
```diff
@@ -0,0 +1,18 @@
+# This migration comes from decidim_debates (originally 20170118141619)
+class CreateDebates < ActiveRecord::Migration[5.0]
```
Comment (oriolgual) [inline db/migrate/20170127093544_create_debates.decidim_debates.rb]: Remove this file, it belongs to the engine.

File: engines/decidim-debates/decidim-debates.gemspec [hunk 13]
```diff
@@ -0,0 +1,22 @@
...
+$LOAD_PATH.push File.expand_path("../lib", __FILE__)
+
+# Describe your gem and declare its dependencies:
+Gem::Specification.new do |s|
+  s.name        = "decidim-debates"
+  s.summary     = "A debates component for decidim's participatory processes."
+  s.description = s.summary
+  s.version     = "0.0.1"
+  s.authors     = ["Josep Jaume Rey Peroy", "Marc Riera Casals", "Oriol Gual Oliva", "Genís Matutes Pujol"]
+  s.email       = ["josepjaume@gmail.com", "mrc2407@gmail.com", "oriolgual@gmail.com", "genis.matutes@gmail.com"]
```
Comment (oriolgual) [inline engines/decidim-debates/decidim-debates.gemspec]: Version, authors & email aren't necessary

File: engines/decidim-debates/decidim-debates.gemspec [hunk 14]
```diff
@@ -0,0 +1,22 @@
...
+
+  s.files = Dir["{app,config,db,lib}/**/*", "Rakefile", "README.md"]
+
+  s.add_dependency "decidim-core"
```
Comment (oriolgual) [inline engines/decidim-debates/decidim-debates.gemspec]: s.add_dependency "decidim-core", Decidim.version

Comment (oriolgual) [inline engines/decidim-debates/decidim-debates.gemspec]: Add comments too:
 s.add_dependency "decidim-comments", Decidim.version

File: engines/decidim-debates/spec/forms/debate_form_spec.rb [hunk 15]
```diff
@@ -0,0 +1,133 @@
+# frozen_literal_string: true
```
Comment (oriolgual) [inline engines/decidim-debates/spec/forms/debate_form_spec.rb]: One of this things is not like the other

File: engines/decidim-debates/decidim-debates.gemspec (hunk 13 above)
Comment (Dor3nz) [inline engines/decidim-debates/decidim-debates.gemspec]: That's actually, wrong? Can't remember when but I was forced to put them in to use the engine.

File: engines/decidim-debates/app/forms/decidim/debates/admin/debate_form.rb (hunk 3 above)
Comment (Dor3nz) [inline engines/decidim-debates/app/forms/decidim/debates/admin/debate_form.rb]: They have no scope, this has to be removed.

File: engines/decidim-debates/decidim-debates.gemspec (hunk 13 above)
Comment (oriolgual) [inline engines/decidim-debates/decidim-debates.gemspec]: It's not wrong but we've omitted these so it "inherits" it from the main gem, which is the only one published.

File: engines/decidim-debates/app/views/decidim/debates/debates/show.html.erb (hunk 9 above)
Comment (Dor3nz) [inline engines/decidim-debates/app/views/decidim/debates/debates/show.html.erb]: Removed author stuff, gotta leave the class for styles.

**Pull request merged**